Enh/hdf5comp

oruebel opened the pull request:
## Motivation

Enable users to  directly specify advanced I/O options for the writing of datasets, specifically chunking and compression. 

# Summary of changes
* Expanded H5DataIO to enable users to specify more advanced I/O settings
* Updated the HDF5IO backend to support the use of the expanded I/O settings of H5DataIO
* Removed direct support Iterable from H5DataIO.write_dataset and instead require users to explicitly wrap iterables using DataChunkIterator. This is to avoid implicit behavior. For convenience I added a factory function for DataChunkIterator to generate it from an Iterable. 
* Added max_shape and dtype functions to AbstractDataChunkIterator to ensure we only assume functionality that is on the base class, rather than DataChunkIterator.
* Added/expanded tests to test the new options

## How to test the behavior?

The following is a direct test derived from the test suit that uses the FORM API only. 

```
import tempfile
import numpy as np
from pynwb.form.backends.hdf5.h5tools import HDF5IO
from pynwb.form.backends.hdf5 import H5DataIO
from pynwb.form.build import DatasetBuilder
test_temp_file = tempfile.NamedTemporaryFile()
test_temp_file.close()
io = HDF5IO(test_temp_file.name)
f = io._file
a = H5DataIO(np.arange(30).reshape(5, 2, 3), compression='gzip', compression_opts=5, shuffle=True, fletcher32=True)
io.write_dataset(f, DatasetBuilder('test_dataset', a, attributes={}))
dset = f['test_dataset']
print("Compression: %s" % str(dset.compression))
print("Compression Options: %s" % str(dset.compression_opts))
print("Shuffle: %s" % str(dset.shuffle))
print("Fletcher32: %s" %  str(dset.fletcher32))
io.close()
```

When using PyNWB containers one woulld hand the H5DataIO object as an input for a specific dataset, e.g., the data of a TimeSeries.

## Checklist

- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR description clearly describes problem and the solution?
- [X] Is your contribution compliant with our coding style ? This can be checked running `flake8` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [X] Have you included the relevant issue number using `#XXX` notation where `XXX` is the issue number ?

oruebel committed:
Added max_shape and dtype methods to the ChunkIterator classes and added corresponding tests

oruebel committed:
Updated H5DataIO to support broader set of I/O options for datasets

oruebel committed:
Added support expanded HDFDataIO options. Removed direct support for Iterable type as part of data write, i.e., require users to explicitly wrap these rather than doing implicit operaitons on the backend

oruebel committed:
Updated/added cases to test advanced I/O options

oruebel committed:
Statisfy flake 8

oruebel committed:
Updated h5utils to fix Python 2 incompatible syntax

pep8speaks commented:
Hello @oruebel! Thanks for updating the PR.



 - In the file [`src/pynwb/form/backends/hdf5/h5tools.py`](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/backends/hdf5/h5tools.py), following are the PEP8 issues :

> [Line 12:22](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/backends/hdf5/h5tools.py#L12): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 17:23](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/backends/hdf5/h5tools.py#L17): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 643:29](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/backends/hdf5/h5tools.py#L643): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 647:48](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/backends/hdf5/h5tools.py#L647): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator



 - In the file [`src/pynwb/form/data_utils.py`](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/data_utils.py), following are the PEP8 issues :

> [Line 108:34](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/data_utils.py#L108): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 176:89](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/data_utils.py#L176): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 372:64](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/data_utils.py#L372): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 375:64](https://github.com/NeurodataWithoutBorders/pynwb/blob/d0851a5d9dd1de0518b345a33fcc94122f987493/src/pynwb/form/data_utils.py#L375): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator







##### Comment last updated on April 10, 2018 at 14:21 Hours UTC

oruebel committed:
Satisfy flake8 on unit tests

oruebel committed:
Added test case to cover the use of recommended chunks from DataChunkIterator in H5DataIO

oruebel committed:
Added warning in case that a compression method other than gzip is being used. Added test case to cover this behavior

oruebel committed:
Updated tests of compression option warning

oruebel committed:
Pass through kwargs from FORMIO.write to FORMIO.write_builder

oruebel committed:
Allow customization of how h5py.Dataset objects are treated on write

- Added link_data parameter to allow setting of the default behavior of HDF5IO when an h5py.Dataset is given as input for write_dataset
- Added link_data parameter to H5DataIO to allow customization of link vs. copy behavior for h5py.Dataset objects

oruebel committed:
Added checks to warn on collision of IO parameters for H5DataIO

oruebel committed:
Satisfy flake 8

oruebel commented:
@ajtritt can you have a look at this. I think this should be ready to be merged now. This mainly updates H5DataIO and HDF5IO to be able to define IO options for datasets, e.g., chunking and compression, but also linking/copy of h5py.Dataset. For the latter I also modified FORMIO to be able to pass parameters from FROMIO.write to FORMIO.write_builder.

bendichter commented:
@jcfr had some good ideas about using sha tags to ensure that linked files are correct.

ajtritt committed:
Merge branch 'dev' into enh/hdf5comp

ajtritt committed:
Merge branch 'dev' into enh/hdf5comp
